File: Introibo/Data/Search/SearchExtractors.swift
import Foundation

// MARK: - SearchExtractors (Phase 1: index core)
//
// Mirror of:
//   android/app/src/main/java/com/lampstandhq/introibo/data/search/SearchExtractors.kt
//
// One extractor per content type. Each turns a slice of ContentStore's
// in-memory content into [SearchDocument]. Extractors are registered in
// SearchIndex.build so adding a content type is:
//   1. add a ContentType case,
//   2. add an extractor,
//   3. register it in SearchIndex.build.
//
// LANGUAGE MODEL (read SearchVernacular.swift for the full story):
//   Content is Latin (`lat`, constant) + ONE user-selected vernacular shown at
//   a time, but the INDEX collects Latin + EVERY available vernacular field so
//   search works for any current/future selection. Extractors NEVER hardcode
//   ".eng" as the only vernacular: structured fragments conform to
//   `Translatable` and expose `translatableFields` (lat + each vernacular in
//   `SearchVernacular.keys`); prose-only English fields are collected through
//   `vernacularProse(...)` which is likewise keyed off `SearchVernacular.keys`.
//   Adding a language is a one-line edit in SearchVernacular.keys plus teaching
//   each model's `vernaculars` map — no extractor edits.

enum SearchExtractors {

    // MARK: Generic translation-text collector

    /// Concatenates an arbitrary list of (optional) translation/text fields
    /// into one whitespace-joined blob, dropping nils/empties. The single
    /// choke point that flattens `lat` + every vernacular field into one
    /// folded blob.
    static func collectText(_ fields: [String?]) -> String {
        fields
            .compactMap { $0 }
            .map { $0.strippingEm }
            .filter { !$0.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty }
            .joined(separator: " ")
    }

    /// Convenience overload for variadic call sites (used for Latin-only or
    /// already-flat field groups).
    static func collectText(_ fields: String?...) -> String {
        collectText(fields)
    }

    /// Collects a per-language prose field that is NOT part of a structured
    /// `Translatable` — e.g. a Saint identity quote or a Reference summary that
    /// today exists only in English. Caller supplies a closure mapping a
    /// vernacular key to that language's value; we iterate `SearchVernacular`
    /// so when an `es`/`fr` prose field is added the same call picks it up.
    static func vernacularProse(_ value: (String) -> String?) -> [String?] {
        SearchVernacular.keys.map { value($0) }
    }

    /// Builds the canonical document id string.
    private static func docID(_ type: ContentType, _ contentID: String, _ position: String? = nil) -> String {
        if let position { return "\(type.rawValue):\(contentID)#\(position)" }
        return "\(type.rawValue):\(contentID)"
    }

    // MARK: - Prayer → 1 doc per prayer

    static func prayers(_ items: [Prayer]) -> [SearchDocument] {
        items.map { p in
            var fields: [String?] = [p.title, p.note]
            fields += vernacularProse { p.vernacular($0) }   // prayer-level body
            for line in p.lines { fields += line.translatableFields }
            // Snippet uses the currently-shipping vernacular (eng) for display.
            let display = collectText([p.eng] + p.lines.map { $0.eng })
            return SearchDocument(
                id: docID(.prayer, p.slug),
                type: .prayer,
                title: p.title,
                subtitle: p.category,
                displayText: display,
                searchText: SearchNormalizer.fold(collectText(fields)),
                target: DeepLinkTarget(type: .prayer, id: p.slug, position: nil)
            )
        }
    }

    // MARK: - MissalSection (Ordinary) → 1 doc per section

    static func missalSections(_ items: [MissalSection]) -> [SearchDocument] {
        items.map { s in
            var fields: [String?] = [s.title, s.label]
            fields += vernacularProse { s.vernacular($0) }   // section heading text
            for line in s.body {
                fields += line.translatableFields
                fields.append(line.rubric)
            }
            let display = collectText([s.english] + s.body.map { $0.eng })
            return SearchDocument(
                id: docID(.missal, s.slug),
                type: .missal,
                title: s.title,
                subtitle: s.label ?? s.english,
                displayText: display,
                searchText: SearchNormalizer.fold(collectText(fields)),
                target: DeepLinkTarget(type: .missal, id: s.slug, position: nil)
            )
        }
    }

    // MARK: - MassProper → 1 doc per element + 1 "feast" title doc

    /// The ordered list of named proper-element fragments on a MassProper.
    /// Each fragment is `Translatable`, so its `translatableFields` already
    /// covers lat + every vernacular. Adding a proper element = add a row here.
    private static func properElements(_ mp: MassProper) -> [(name: String, fields: [String?])] {
        func fields(_ t: Translatable?) -> [String?] { t?.translatableFields ?? [] }
        return [
            ("introit",       fields(mp.introit)),
            ("collect",       fields(mp.collect)),
            ("epistle",       fields(mp.epistle)),
            ("gradual",       fields(mp.gradual)),
            ("alleluia",      fields(mp.alleluia)),
            ("tract",         fields(mp.tract)),
            ("sequence",      fields(mp.sequence)),
            ("gospel",        fields(mp.gospel)),
            ("offertory",     fields(mp.offertory)),
            ("secret",        fields(mp.secret)),
            ("communion",     fields(mp.communion)),
            ("postcommunion", fields(mp.postcommunion)),
        ]
    }

    static func propers(_ items: [MassProper]) -> [SearchDocument] {
        var docs: [SearchDocument] = []
        for mp in items {
            // 1 title-only "feast" doc for calendar findability. Title is Latin,
            // english is the vernacular; collected generically.
            var feastFields: [String?] = [mp.title]
            feastFields += vernacularProse { mp.vernacular($0) }
            let feastSearch = SearchNormalizer.fold(collectText(feastFields))
            if !feastSearch.isEmpty {
                docs.append(SearchDocument(
                    id: docID(.missal, mp.slug, "feast"),
                    type: .missal,
                    title: mp.title,
                    subtitle: mp.english,
                    displayText: collectText(mp.english),
                    searchText: feastSearch,
                    target: DeepLinkTarget(type: .missal, id: mp.slug, position: "feast")
                ))
            }
            // 1 doc per element that has text.
            for element in properElements(mp) {
                let folded = SearchNormalizer.fold(collectText(element.fields))
                if folded.isEmpty { continue }
                docs.append(SearchDocument(
                    id: docID(.missal, mp.slug, element.name),
                    type: .missal,
                    title: mp.title,
                    subtitle: element.name,
                    displayText: collectText(element.fields),
                    searchText: folded,
                    target: DeepLinkTarget(type: .missal, id: mp.slug, position: element.name)
                ))
            }
        }
        return docs
    }

    // MARK: - Hour (template hours) → 1 doc per Part that has text

    /// Every translatable field on an Hour.Part, collected lat + vernacular via
    /// the part's own `translatableFields` plus its flattened verses.
    private static func partFields(_ part: Hour.Part) -> [String?] {
        var f: [String?] = [part.label, part.title]
        f += part.translatableFields
        if let verses = part.verses {
            for v in verses { f += v.translatableFields }
        }
        return f
    }

    static func hours(_ items: [Hour]) -> [SearchDocument] {
        var docs: [SearchDocument] = []
        for hour in items {
            for (i, part) in hour.parts.enumerated() {
                let fields = partFields(part)
                let folded = SearchNormalizer.fold(collectText(fields))
                if folded.isEmpty { continue }
                let partTitle = part.title ?? part.label ?? hour.eng
                docs.append(SearchDocument(
                    id: docID(.office, hour.slug, "part:\(i)"),
                    type: .office,
                    title: hour.eng,
                    subtitle: partTitle,
                    displayText: collectText(fields),
                    searchText: folded,
                    target: DeepLinkTarget(type: .office, id: hour.slug, position: "part:\(i)")
                ))
            }
        }
        return docs
    }

    // MARK: - ReferenceEntry → reference (1/entry) or calendar (1/season)

    private static let calendarCategory = "Calendarium"

    static func reference(_ items: [ReferenceEntry]) -> [SearchDocument] {
        items
            .filter { $0.cat != calendarCategory }
            .map { e in
                var fields: [String?] = [e.title, e.latin]
                // Prose body (summary/history/practice/notes) is per-vernacular.
                fields += vernacularProse { e.vernacularProse($0) }
                fields += (e.scripture?.translatableFields ?? [])
                return SearchDocument(
                    id: docID(.reference, e.slug),
                    type: .reference,
                    title: e.title,
                    subtitle: e.cat,
                    displayText: collectText(e.summary),
                    searchText: SearchNormalizer.fold(collectText(fields)),
                    target: DeepLinkTarget(type: .reference, id: e.slug, position: nil)
                )
            }
    }

    static func calendar(_ items: [ReferenceEntry]) -> [SearchDocument] {
        items
            .filter { $0.cat == calendarCategory }
            .map { e in
                var fields: [String?] = [e.title, e.latin]
                fields += vernacularProse { e.vernacularProse($0) }
                return SearchDocument(
                    id: docID(.calendar, e.slug),
                    type: .calendar,
                    title: e.title,
                    subtitle: e.latin ?? e.cat,
                    displayText: collectText(e.summary),
                    searchText: SearchNormalizer.fold(collectText(fields)),
                    target: DeepLinkTarget(type: .calendar, id: e.slug, position: nil)
                )
            }
    }

    // MARK: - Saint → 1 identity + 1 per section + 1 per prayer

    static func saints(_ items: [Saint]) -> [SearchDocument] {
        var docs: [SearchDocument] = []
        for s in items {
            // Identity doc. Name is language-neutral; quote/penance are prose.
            var idFields: [String?] = [s.name, s.title, s.penanceLatin]
            idFields += vernacularProse { s.vernacularProse($0) }
            docs.append(SearchDocument(
                id: docID(.saint, s.slug),
                type: .saint,
                title: s.name,
                subtitle: s.title,
                displayText: collectText(s.quote),
                searchText: SearchNormalizer.fold(collectText(idFields)),
                target: DeepLinkTarget(type: .saint, id: s.slug, position: nil)
            ))
            // Section docs.
            for (i, section) in s.sections.enumerated() {
                var fields: [String?] = section.translatableFields
                for pr in section.practices { fields.append(pr.t); fields.append(pr.d) }
                let folded = SearchNormalizer.fold(collectText(fields))
                if folded.isEmpty { continue }
                docs.append(SearchDocument(
                    id: docID(.saint, s.slug, "section:\(i)"),
                    type: .saint,
                    title: s.name,
                    subtitle: s.title,
                    displayText: collectText(section.eng),
                    searchText: folded,
                    target: DeepLinkTarget(type: .saint, id: s.slug, position: "section:\(i)")
                ))
            }
            // Prayer docs.
            for (i, pr) in (s.prayers ?? []).enumerated() {
                var fields: [String?] = [pr.title, pr.latin, pr.note]
                fields += vernacularProse { pr.vernacular($0) }
                let folded = SearchNormalizer.fold(collectText(fields))
                if folded.isEmpty { continue }
                docs.append(SearchDocument(
                    id: docID(.saint, s.slug, "prayer:\(i)"),
                    type: .saint,
                    title: pr.title,
                    subtitle: s.name,
                    displayText: collectText(pr.eng),
                    searchText: folded,
                    target: DeepLinkTarget(type: .saint, id: s.slug, position: "prayer:\(i)")
                ))
            }
        }
        return docs
    }
}

// MARK: - Translatable conformances
//
// The ONLY place a vernacular field name appears. When the content model gains
// `es`/`fr`/… fields, add them to `vernaculars` (and the prose accessors) here
// and to `SearchVernacular.keys`; nothing in the extractors above changes.

extension ProperText: Translatable {
    var latText: String? { lat }
    var vernaculars: [String: String?] { ["eng": eng] }
}

extension ProperReading: Translatable {
    var latText: String? { lat }
    var vernaculars: [String: String?] { ["eng": eng] }
}

extension Prayer.Line: Translatable {
    var latText: String? { lat }
    var vernaculars: [String: String?] { ["eng": eng] }
}

extension MissalSection.Line: Translatable {
    var latText: String? { lat }
    var vernaculars: [String: String?] { ["eng": eng] }
}

extension Hour.Part.Verse: Translatable {
    var latText: String? { lat }
    var vernaculars: [String: String?] { ["eng": eng] }
}

extension ReferenceEntry.Scripture: Translatable {
    var latText: String? { lat }
    var vernaculars: [String: String?] { ["eng": eng] }
}

extension Saint.Section: Translatable {
    var latText: String? { lat }
    var vernaculars: [String: String?] { ["eng": eng] }
}

// Hour.Part carries many parallel lat/eng pairs; expose them all so a single
// `translatableFields` yields every paired field for both languages.
extension Hour.Part: Translatable {
    var latText: String? {
        SearchExtractors.collectText(lat, latR, v1Lat, r1Lat, v2Lat, r2Lat, antiphonLat)
    }
    var vernaculars: [String: String?] {
        ["eng": SearchExtractors.collectText(eng, engR, v1Eng, r1Eng, v2Eng, r2Eng, engBody, antiphonEng)]
    }
}

// MARK: - Prose-only vernacular accessors
//
// These models currently hold their vernacular as flat English prose (no Latin
// counterpart per field). Each returns the value for a given vernacular key so
// the extractor stays generic; add `es`/`fr` cases here alongside the model
// fields when those languages land.

extension Prayer {
    func vernacular(_ key: String) -> String? {
        switch key {
        case "eng": return eng
        default: return nil
        }
    }
}

extension MissalSection {
    func vernacular(_ key: String) -> String? {
        switch key {
        case "eng": return english
        default: return nil
        }
    }
}

extension MassProper {
    /// The feast's vernacular display name (`english` today).
    func vernacular(_ key: String) -> String? {
        switch key {
        case "eng": return english
        default: return nil
        }
    }
}

extension ReferenceEntry {
    /// All prose body fields for a vernacular, joined. English-only today.
    func vernacularProse(_ key: String) -> String? {
        switch key {
        case "eng": return SearchExtractors.collectText(summary, history, practice, notes)
        default: return nil
        }
    }
}

extension Saint {
    /// Identity-level prose (quote + penance) for a vernacular. English today.
    func vernacularProse(_ key: String) -> String? {
        switch key {
        case "eng": return SearchExtractors.collectText(quote, penance)
        default: return nil
        }
    }
}

extension Saint.SaintPrayer {
    func vernacular(_ key: String) -> String? {
        switch key {
        case "eng": return eng
        default: return nil
        }
    }
}

File: android/app/src/main/java/com/lampstandhq/introibo/data/search/SearchExtractors.kt
package com.lampstandhq.introibo.data.search

import com.lampstandhq.introibo.data.model.Hour
import com.lampstandhq.introibo.data.model.MassProper
import com.lampstandhq.introibo.data.model.MissalSection
import com.lampstandhq.introibo.data.model.Prayer
import com.lampstandhq.introibo.data.model.ProperReading
import com.lampstandhq.introibo.data.model.ProperText
import com.lampstandhq.introibo.data.model.ReferenceEntry
import com.lampstandhq.introibo.data.model.Saint
import com.lampstandhq.introibo.data.model.strippingEm

// MARK: - SearchExtractors (Phase 1: index core)
//
// Mirror of:
//   Introibo/Data/Search/SearchExtractors.swift
//
// One extractor per content type. Each turns a slice of ContentStore's
// in-memory content into List<SearchDocument>. Extractors are run (and
// partitioned) by SearchIndex.build, so adding a content type is:
//   1. add a ContentType case,
//   2. add an extractor,
//   3. partition it in SearchIndex.build.
//
// LANGUAGE MODEL (read SearchVernacular.kt for the full story):
//   Content is Latin (`lat`, constant) + ONE user-selected vernacular shown at
//   a time, but the INDEX collects Latin + EVERY available vernacular field so
//   search works for any current/future selection. Extractors NEVER hardcode
//   `eng` as the only vernacular: structured fragments are flattened via
//   `translatable(lat, vernaculars)` (lat + each vernacular in
//   `SearchVernacular.keys`); prose-only English fields are collected through
//   `vernacularProse(...)` which is likewise keyed off `SearchVernacular.keys`.
//   Adding a language is a one-line edit in SearchVernacular.keys plus teaching
//   each model's vernacular map below — no extractor-body edits.

object SearchExtractors {

    // ---- Generic translation-text collector ----

    /**
     * Concatenates an arbitrary list of nullable translation/text fields into
     * one whitespace-joined blob, dropping nulls/empties. The single choke
     * point that flattens `lat` + every vernacular field into one folded blob.
     */
    fun collectText(fields: List<String?>): String =
        fields
            .filterNotNull()
            .map { it.strippingEm }
            .filter { it.isNotBlank() }
            .joinToString(" ")

    /** Convenience vararg overload (Latin-only or already-flat field groups). */
    fun collectText(vararg fields: String?): String = collectText(fields.toList())

    // ---- Translatable flattening (mirror of iOS `Translatable.translatableFields`) ----
    //
    // Kotlin can't retroactively make the model data classes implement an
    // interface, so instead of `fragment.translatableFields` we flatten a
    // (latText, vernaculars-map) pair here. Iterating `SearchVernacular.keys`
    // (not the map) keeps order stable and auto-includes future languages.

    private fun translatable(latText: String?, vernaculars: Map<String, String?>): List<String?> {
        val out = mutableListOf<String?>(latText)
        for (key in SearchVernacular.keys) {
            if (vernaculars.containsKey(key)) out.add(vernaculars[key])
        }
        return out
    }

    /**
     * Collects a per-language prose field that is NOT part of a structured
     * fragment — e.g. a Saint identity quote or a Reference summary that today
     * exists only in English. Caller supplies a lambda mapping a vernacular key
     * to that language's value; we iterate `SearchVernacular.keys` so an added
     * `es`/`fr` prose field is picked up by the same call.
     */
    private fun vernacularProse(value: (String) -> String?): List<String?> =
        SearchVernacular.keys.map { value(it) }

    // ---- Per-model translatable adapters ----
    // The ONLY place a vernacular field name ("eng") appears. Add `es`/`fr`
    // entries here (and to SearchVernacular.keys) when those languages land;
    // nothing in the extractor bodies below changes.

    private fun fields(t: ProperText?): List<String?> =
        if (t == null) emptyList() else translatable(t.lat, mapOf("eng" to t.eng))

    private fun fields(r: ProperReading?): List<String?> =
        if (r == null) emptyList() else translatable(r.lat, mapOf("eng" to r.eng))

    private fun fields(l: Prayer.Line): List<String?> =
        translatable(l.lat, mapOf("eng" to l.eng))

    private fun fields(l: MissalSection.Line): List<String?> =
        translatable(l.lat, mapOf("eng" to l.eng))

    private fun fields(v: Hour.Part.Verse): List<String?> =
        translatable(v.lat, mapOf("eng" to v.eng))

    private fun fields(s: ReferenceEntry.Scripture?): List<String?> =
        if (s == null) emptyList() else translatable(s.lat, mapOf("eng" to s.eng))

    private fun fields(sec: Saint.Section): List<String?> =
        translatable(sec.lat, mapOf("eng" to sec.eng))

    /** Hour.Part carries many parallel lat/eng pairs; expose them all at once. */
    private fun fields(part: Hour.Part): List<String?> = translatable(
        latText = collectText(
            part.lat, part.latR, part.v1Lat, part.r1Lat, part.v2Lat, part.r2Lat, part.antiphonLat
        ),
        vernaculars = mapOf(
            "eng" to collectText(
                part.eng, part.engR, part.v1Eng, part.r1Eng, part.v2Eng, part.r2Eng,
                part.engBody, part.antiphonEng
            )
        ),
    )

    // ---- Prose-only vernacular accessors (English-only today) ----

    private fun prayerVernacular(p: Prayer, key: String): String? =
        if (key == "eng") p.eng else null

    private fun missalVernacular(s: MissalSection, key: String): String? =
        if (key == "eng") s.english else null

    private fun properVernacular(mp: MassProper, key: String): String? =
        if (key == "eng") mp.english else null

    private fun referenceProse(e: ReferenceEntry, key: String): String? =
        if (key == "eng") collectText(e.summary, e.history, e.practice, e.notes) else null

    private fun saintProse(s: Saint, key: String): String? =
        if (key == "eng") collectText(s.quote, s.penance) else null

    private fun saintPrayerVernacular(pr: Saint.SaintPrayer, key: String): String? =
        if (key == "eng") pr.eng else null

    /** Builds the canonical document id string. */
    private fun docId(type: ContentType, contentId: String, position: String? = null): String =
        if (position != null) "${type.wire}:$contentId#$position" else "${type.wire}:$contentId"

    // ---- Prayer → 1 doc per prayer ----

    fun prayers(items: List<Prayer>): List<SearchDocument> = items.map { p ->
        val f = mutableListOf<String?>(p.title, p.note)
        f += vernacularProse { prayerVernacular(p, it) }   // prayer-level body
        for (line in p.lines) f += fields(line)
        // Snippet uses the currently-shipping vernacular (eng) for display.
        val display = collectText(listOf(p.eng) + p.lines.map { it.eng })
        SearchDocument(
            id = docId(ContentType.PRAYER, p.slug),
            type = ContentType.PRAYER,
            title = p.title,
            subtitle = p.category,
            displayText = display,
            searchText = SearchNormalizer.fold(collectText(f)),
            target = DeepLinkTarget(ContentType.PRAYER, p.slug, null),
        )
    }

    // ---- MissalSection (Ordinary) → 1 doc per section ----

    fun missalSections(items: List<MissalSection>): List<SearchDocument> = items.map { s ->
        val f = mutableListOf<String?>(s.title, s.label)
        f += vernacularProse { missalVernacular(s, it) }   // section heading text
        for (line in s.body) {
            f += fields(line)
            f.add(line.rubric)
        }
        val display = collectText(listOf(s.english) + s.body.map { it.eng })
        SearchDocument(
            id = docId(ContentType.MISSAL, s.slug),
            type = ContentType.MISSAL,
            title = s.title,
            subtitle = s.label ?: s.english,
            displayText = display,
            searchText = SearchNormalizer.fold(collectText(f)),
            target = DeepLinkTarget(ContentType.MISSAL, s.slug, null),
        )
    }

    // ---- MassProper → 1 doc per element + 1 "feast" title doc ----

    /**
     * The ordered list of named proper-element fragments on a MassProper. Each
     * fragment's `fields(...)` already covers lat + every vernacular. Adding a
     * proper element = add a row here.
     */
    private fun properElements(mp: MassProper): List<Pair<String, List<String?>>> = listOf(
        "introit" to fields(mp.introit),
        "collect" to fields(mp.collect),
        "epistle" to fields(mp.epistle),
        "gradual" to fields(mp.gradual),
        "alleluia" to fields(mp.alleluia),
        "tract" to fields(mp.tract),
        "sequence" to fields(mp.sequence),
        "gospel" to fields(mp.gospel),
        "offertory" to fields(mp.offertory),
        "secret" to fields(mp.secret),
        "communion" to fields(mp.communion),
        "postcommunion" to fields(mp.postcommunion),
    )

    fun propers(items: List<MassProper>): List<SearchDocument> {
        val docs = mutableListOf<SearchDocument>()
        for (mp in items) {
            // 1 title-only "feast" doc for calendar findability.
            val feastFields = mutableListOf<String?>(mp.title)
            feastFields += vernacularProse { properVernacular(mp, it) }
            val feastSearch = SearchNormalizer.fold(collectText(feastFields))
            if (feastSearch.isNotEmpty()) {
                docs.add(
                    SearchDocument(
                        id = docId(ContentType.MISSAL, mp.slug, "feast"),
                        type = ContentType.MISSAL,
                        title = mp.title,
                        subtitle = mp.english,
                        displayText = collectText(mp.english),
                        searchText = feastSearch,
                        target = DeepLinkTarget(ContentType.MISSAL, mp.slug, "feast"),
                    )
                )
            }
            // 1 doc per element that has text.
            for ((name, elementFields) in properElements(mp)) {
                val folded = SearchNormalizer.fold(collectText(elementFields))
                if (folded.isEmpty()) continue
                docs.add(
                    SearchDocument(
                        id = docId(ContentType.MISSAL, mp.slug, name),
                        type = ContentType.MISSAL,
                        title = mp.title,
                        subtitle = name,
                        displayText = collectText(elementFields),
                        searchText = folded,
                        target = DeepLinkTarget(ContentType.MISSAL, mp.slug, name),
                    )
                )
            }
        }
        return docs
    }

    // ---- Hour (template hours) → 1 doc per Part that has text ----

    private fun partFields(part: Hour.Part): List<String?> {
        val f = mutableListOf<String?>(part.label, part.title)
        f += fields(part)
        part.verses?.forEach { v -> f += fields(v) }
        return f
    }

    fun hours(items: List<Hour>): List<SearchDocument> {
        val docs = mutableListOf<SearchDocument>()
        for (hour in items) {
            hour.parts.forEachIndexed { i, part ->
                val f = partFields(part)
                val folded = SearchNormalizer.fold(collectText(f))
                if (folded.isEmpty()) return@forEachIndexed
                val partTitle = part.title ?: part.label ?: hour.eng
                docs.add(
                    SearchDocument(
                        id = docId(ContentType.OFFICE, hour.slug, "part:$i"),
                        type = ContentType.OFFICE,
                        title = hour.eng,
                        subtitle = partTitle,
                        displayText = collectText(f),
                        searchText = folded,
                        target = DeepLinkTarget(ContentType.OFFICE, hour.slug, "part:$i"),
                    )
                )
            }
        }
        return docs
    }

    // ---- ReferenceEntry → reference (1/entry) or calendar (1/season) ----

    private const val CALENDAR_CATEGORY = "Calendarium"

    fun reference(items: List<ReferenceEntry>): List<SearchDocument> =
        items.filter { it.cat != CALENDAR_CATEGORY }.map { e ->
            val f = mutableListOf<String?>(e.title, e.latin)
            f += vernacularProse { referenceProse(e, it) }
            f += fields(e.scripture)
            SearchDocument(
                id = docId(ContentType.REFERENCE, e.slug),
                type = ContentType.REFERENCE,
                title = e.title,
                subtitle = e.cat,
                displayText = collectText(e.summary),
                searchText = SearchNormalizer.fold(collectText(f)),
                target = DeepLinkTarget(ContentType.REFERENCE, e.slug, null),
            )
        }

    fun calendar(items: List<ReferenceEntry>): List<SearchDocument> =
        items.filter { it.cat == CALENDAR_CATEGORY }.map { e ->
            val f = mutableListOf<String?>(e.title, e.latin)
            f += vernacularProse { referenceProse(e, it) }
            SearchDocument(
                id = docId(ContentType.CALENDAR, e.slug),
                type = ContentType.CALENDAR,
                title = e.title,
                subtitle = e.latin ?: e.cat,
                displayText = collectText(e.summary),
                searchText = SearchNormalizer.fold(collectText(f)),
                target = DeepLinkTarget(ContentType.CALENDAR, e.slug, null),
            )
        }

    // ---- Saint → 1 identity + 1 per section + 1 per prayer ----

    fun saints(items: List<Saint>): List<SearchDocument> {
        val docs = mutableListOf<SearchDocument>()
        for (s in items) {
            // Identity doc. Name is language-neutral; quote/penance are prose.
            val idFields = mutableListOf<String?>(s.name, s.title, s.penanceLatin)
            idFields += vernacularProse { saintProse(s, it) }
            docs.add(
                SearchDocument(
                    id = docId(ContentType.SAINT, s.slug),
                    type = ContentType.SAINT,
                    title = s.name,
                    subtitle = s.title,
                    displayText = collectText(s.quote),
                    searchText = SearchNormalizer.fold(collectText(idFields)),
                    target = DeepLinkTarget(ContentType.SAINT, s.slug, null),
                )
            )
            // Section docs.
            s.sections.forEachIndexed { i, section ->
                val f = fields(section).toMutableList()
                for (pr in section.practices) { f.add(pr.t); f.add(pr.d) }
                val folded = SearchNormalizer.fold(collectText(f))
                if (folded.isEmpty()) return@forEachIndexed
                docs.add(
                    SearchDocument(
                        id = docId(ContentType.SAINT, s.slug, "section:$i"),
                        type = ContentType.SAINT,
                        title = s.name,
                        subtitle = s.title,
                        displayText = collectText(section.eng),
                        searchText = folded,
                        target = DeepLinkTarget(ContentType.SAINT, s.slug, "section:$i"),
                    )
                )
            }
            // Prayer docs.
            (s.prayers ?: emptyList()).forEachIndexed { i, pr ->
                val f = mutableListOf<String?>(pr.title, pr.latin, pr.note)
                f += vernacularProse { saintPrayerVernacular(pr, it) }
                val folded = SearchNormalizer.fold(collectText(f))
                if (folded.isEmpty()) return@forEachIndexed
                docs.add(
                    SearchDocument(
                        id = docId(ContentType.SAINT, s.slug, "prayer:$i"),
                        type = ContentType.SAINT,
                        title = pr.title,
                        subtitle = s.name,
                        displayText = collectText(pr.eng),
                        searchText = folded,
                        target = DeepLinkTarget(ContentType.SAINT, s.slug, "prayer:$i"),
                    )
                )
            }
        }
        return docs
    }
}
